typing/typemod.ml

```diff
@@ -143,37 +143,56 @@ let type_open_ ?used_slot ?toplevel ovf env loc lid =
       ignore (extract_sig_open env lid.loc md.md_type);
       assert false
 
-let type_initially_opened_module env module_name =
-  let loc = Location.in_file "compiler internals" in
-  let lid = { Asttypes.loc; txt = Longident.Lident module_name } in
-  let path = Typetexp.lookup_module ~load:true env lid.loc lid.txt in
-  match Env.open_signature_of_initially_opened_module path env with
-  | Some env -> path, env
-  | None ->
-      let md = Env.find_module path env in
-      ignore (extract_sig_open env lid.loc md.md_type);
-      assert false
-
 let initial_env ~loc ~safe_string ~initially_opened_module
-      ~open_implicit_modules =
+    ~open_implicit_modules =
   let env =
     if safe_string then
       Env.initial_safe_string
     else
       Env.initial_unsafe_string
   in
-  let env =
-    match initially_opened_module with
-    | None -> env
-    | Some name ->
-      snd (type_initially_opened_module env name)
-  in
-  let open_implicit_module env m =
+  let open_module env m =
     let open Asttypes in
     let lid = {loc; txt = Longident.parse m } in
     snd (type_open_ Override env lid.loc lid)
   in
-  List.fold_left open_implicit_module env open_implicit_modules
+  let add_units env units =
+    String.Set.fold
+      (fun name env ->
+         Env.add_persistent_structure (Ident.create_persistent name) env)
+      units
+      env
+  in
+  let units =
+    List.rev_map Env.persistent_structures_of_dir (Load_path.get ())
+  in
+  let env, units =
+    match initially_opened_module with
+    | None -> (env, units)
+    | Some m ->
+        (* Locate the directory that contains [m], adds the units it
+           contains to the environment and open [m] in the resulting
+           environment. *)
```

Comment:
Can you explain the logic here?  Why don't we add all units from all directories first and then open the initially_opened_module in the resulting env?

Reply:
If we ever make the `initially_opened_module` customizable (currently, it is always "Stdlib" or nothing), the logic would seem a bit ad hoc: modules in the same directory than the initially_opened_module wouldn't be allowed to override sub-modules of that one, but modules in any other -I directory would.  Perhaps not a big deal in practice, though.

> In fact, after this PR we almost want to give meaning to the order in which the various -I and -open flags are passed.

Yes, it would seem quite natural then.  People could explicitly pass the same -I directory several times if needed.